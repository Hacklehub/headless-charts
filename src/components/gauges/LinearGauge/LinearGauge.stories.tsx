import LinearGauge from '.';
import { Meta } from '@storybook/react';

export default {
  title: 'Gauges/LinearGauge',
  component: LinearGauge,
  tags: ['lineargauges'],
} as Meta;

export const Default = {
  args: {
    id: 'linear-gauge-default',
    className: '',
    label: 'Linear Gauge Graph',
    data: 47,
    max: 100,
    gaugeHeight: 6,
    marginLeft: 40,
    marginTop: 30,
    marginRight: 40,
    marginBottom: 10,
  },
};

export const Drawing = {
  args: {
    ...Default.args,
    drawing: { duration: 2000 },
  },
};

export const Error = {
  args: {
    id: 'linear-gauge-with-error',
    label: 'linear Gauge With Error',
    data: 23,
    max: 25,
    error: { data: 0.5, className: 'text-red' },
  },
};

export const ToolTip = {
  args: {
    id: 'linear-gauge-with-tooltip',
    className: '',
    label: 'Linear Gauge Graph With Tooltip',
    data: 67,
    max: 100,
    drawing: { duration: 2000 },
    tooltip: {
      html: (data = 45, error = { data: 20, className: 'bg-red' }) => {
        return (
          <div>
            <p>Data: ${data}</p>${error ? `<p>Error: ${error.data}</p>` : ''}
          </div>
        );
      },
      className: 'bg-gray-800 text-white p-2 rounded',
    },
  },
};
