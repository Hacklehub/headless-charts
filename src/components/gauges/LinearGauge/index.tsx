import React, { useEffect, useRef } from 'react';
/* eslint-disable @typescript-eslint/no-explicit-any */
import { ValueFn, pointer, select, selectAll } from 'd3';

import { interpolate } from 'd3-interpolate';
import { mergeTailwindClasses } from '../../../utils';
import { scaleLinear } from 'd3';

interface LinearGaugeProps {
  id: string;
  className?: string;
  label:
    | string
    | number
    | boolean
    | ValueFn<SVGTextElement, unknown, string | number | boolean | null>
    | null;
  data?: number;
  max?: number;
  error?: { data: number; className?: string };
  gaugeHeight?: number;
  margin?: {
    left?: number;
    right?: number;
    top?: number;
    bottom?: number;
  };
  drawing?: { duration: number };
  tooltip?: {
    html?: string;
    className?: string;
  };
  classNameGauge?: string;
  classNameGaugeBg?: string;
  startAngle?: number;
  nameKey: string;
  // endAngle?: number;
}

const LinearGauge = ({
  id,
  className,
  label,
  data = 0,
  max = 1,
  error,
  gaugeHeight = 6,
  margin = {
    left: 40,
    top: 30,
    right: 40,
    bottom: 10,
  },
  drawing = { duration: 1000 },
  tooltip = undefined,
  classNameGauge = '',
  classNameGaugeBg = '',
  startAngle = 0,
  nameKey = 'name',
}: // endAngle = 360 + startAngle,
LinearGaugeProps) => {
  const setup = React.useCallback(() => {
    const svg = select(`#${id}`);

    svg.selectAll('*').remove();

    const width = +svg.style('width').split('px')[0];
    const height = +svg.style('height').split('px')[0];

    const xFn = scaleLinear()
      .domain([0, max])
      .range([margin.left || 0, width - (margin.right || 0)]);

    const g = svg.append('g');

    const gaugeG = g.append('g');

    gaugeG
      .append('text')
      .attr('class', 'fill-current text-lg')
      .text(label)
      .attr('text-anchor', 'middle')
      .attr(
        'x',
        (margin.left || 0) +
          (width - (margin.left || 0) - (margin.right || 0)) / 2
      )
      .attr('y', height - (margin.top || 0));

    gaugeG
      .append('rect')
      .attr(
        'class',
        mergeTailwindClasses(
          'fill-current stroke-current text-gray-300 dark:text-gray-700',
          classNameGaugeBg
        )
      )
      .attr('x', margin.left || 0)
      .attr('y', height - (margin.bottom || 0) - gaugeHeight)
      .attr('width', width - (margin.left || 0) - (margin.right || 0))
      .attr('height', gaugeHeight)
      .attr('ry', gaugeHeight / 2)
      .on('mouseenter', function () {
        tooltip &&
          tooltipDiv
            .style('opacity', 1)
            .html(
              tooltip?.html || `Data: ${data} <br/> Error: ${error?.data || 0} `
            );
      })
      .on('mousemove', function (event) {
        const [bX, bY] = pointer(event, select('body'));
        tooltipDiv.style('left', `${bX + 10}px`).style('top', `${bY + 10}px`);
      })
      .on('mouseleave', function () {
        tooltip &&
          tooltipDiv
            .style('opacity', '0')
            .style('left', `0px`)
            .style('top', `0px`);
      });

    gaugeG
      .append('rect')
      .attr(
        'class',
        mergeTailwindClasses(
          'data-rect fill-current stroke-current',
          classNameGauge
        )
      )
      .attr('x', margin.left || 0)
      .attr('y', height - (margin.bottom || 0) - gaugeHeight)
      .attr('width', 0)
      .attr('height', gaugeHeight)
      .attr('ry', gaugeHeight / 2)
      .transition()
      .duration(drawing.duration)
      .attr('width', xFn(data) - xFn(0));

    error &&
      error.data &&
      gaugeG
        .append('rect')
        .attr(
          'class',
          mergeTailwindClasses(
            'data-rect fill-current stroke-current text-red-700',
            (error && error.className) || ''
          )
        )
        .attr('x', width - (margin.right || 0))
        .attr('y', height - (margin.bottom || 0) - gaugeHeight)
        .attr('ry', gaugeHeight / 2)
        .attr('width', 0)
        .attr('height', gaugeHeight)
        .transition()
        .duration(drawing.duration)
        .attr('x', () => xFn(max - (error.data || 0)))
        .attr('width', () => {
          return xFn(max) - xFn(max - (error.data || 0));
        });

    const tooltipDiv = select('body')
      .append('div')
      .attr('id', 'tooltip')
      .style('position', 'absolute')
      .style('opacity', '0')
      .attr(
        'class',
        mergeTailwindClasses('tooltip ', tooltip && tooltip.className)
      );
  }, [
    classNameGauge,
    classNameGaugeBg,
    data,
    drawing,
    error,
    gaugeHeight,
    id,
    label,
    margin,
    max,
    tooltip,
  ]);

  const previousArcs = useRef<any[]>([]);

  const refreshChart = React.useCallback(() => {
    const svg = select(`#${id}`);
    const width = +svg.style('width').split('px')[0];
    /* eslint-disable */

    const xFn = scaleLinear()
      .domain([0, max])
      .range([margin.left || 0, width - (margin.right || 0)]);

    select('.data-rect')
      .transition()
      .duration(drawing?.duration || 0)
      .attr('width', xFn(data) - xFn(0))
      .attrTween('d', function (d) {
        const previousArc = previousArcs.current.find(
          // @ts-ignore
          (a) => a.data[nameKey] === d.data[nameKey]
        );

        const i = interpolate(
          {
            startAngle: previousArc?.startAngle || (startAngle / 180) * Math.PI,
            endAngle: previousArc?.endAngle || (startAngle / 180) * Math.PI,
          },
          // @ts-ignore
          d
        );

        // @ts-ignore
        return (t) => arcFn(i(t));
      });

    const timeOut = setTimeout(() => {
      // @ts-ignore
      previousArcs.current = arcs;
    }, drawing?.duration);

    /* eslint-enable */
    return () => {
      clearTimeout(timeOut);
    };
  }, [data, drawing, id, margin, max]);

  useEffect(() => {
    setup();
  }, [setup]);

  useEffect(() => {
    refreshChart();
    return () => {
      selectAll('.tooltip').remove();
    };
  }, [data, max, refreshChart]);

  return (
    <svg
      id={id}
      className={mergeTailwindClasses(
        className,
        `w-full md:w-6/12 lg:w-4/12 dark:bg-gray-800 text-gray-900 dark:text-gray-50 chart `
      )}
    />
  );
};

export default LinearGauge;
