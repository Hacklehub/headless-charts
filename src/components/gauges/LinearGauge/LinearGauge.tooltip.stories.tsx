import { Meta, StoryObj } from '@storybook/react';

import LinearGauge from '.';

const meta: Meta<typeof LinearGauge> = {
  title: 'Gauge/LinearGauge/Tooltips',
  component: LinearGauge,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof LinearGauge>;

export const ToolTip: Story = {
  args: {
    id: 'linear-gauge-with-tooltip',
    className: '',
    label: 'Linear Gauge Graph With Tooltip',
    data: 67,
    max: 100,
    drawing: { duration: 2000 },
    tooltip: {
      className: 'bg-gray-800 text-white p-2 rounded',
    },
  },
};

export const ToolTipWithCustomHtml: Story = {
  args: {
    ...ToolTip.args,
    id: 'linear-gauge-with-tooltip-custom-html',
    error: { data: 2 },
    tooltip: {
      html: `<div class='bg-gray-800 text-white p-2 rounded'>67% with 2% error</div>`,
    },
  },
};
