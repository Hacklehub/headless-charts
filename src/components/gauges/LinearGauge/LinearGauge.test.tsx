import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import LinearGauge from '.';

describe('LinearGauge', () => {
  it('renders LinearGauge component for snapshot test', () => {
    render(<LinearGauge id='linear-gauge'
      className='h-12'
      label='Linear Gauge Graph'
      data={0.47} />);
    const linearGauge = screen.getByTestId('linear-gauge');
    expect(linearGauge).toMatchSnapshot();
  });

  it('renders LinearGauge component', () => {
    render(<LinearGauge id='linear-gauge'
      className='h-12'
      label='Linear Gauge Graph'
      data={0.47} />);
    const linearGauge = screen.getByTestId('linear-gauge');
    expect(linearGauge).toBeInTheDocument();
  });

})
