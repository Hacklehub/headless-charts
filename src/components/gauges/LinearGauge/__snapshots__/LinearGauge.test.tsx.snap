// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LinearGauge > renders LinearGauge component for snapshot test 1`] = `
<svg
  class=" chart dark:bg-gray-800 dark:text-gray-50 h-12 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="linear-gauge"
  id="linear-gauge"
>
  <g>
    <g>
      <text
        class="fill-current text-lg"
        text-anchor="middle"
        x="0"
        y="-30"
      >
        Linear Gauge Graph
      </text>
      <rect
        class="dark:text-gray-700 fill-current stroke-current text-gray-300"
        height="6"
        ry="3"
        width="-80"
        x="40"
        y="-16"
      />
      <rect
        class="data-rect fill-current stroke-current"
        height="6"
        ry="3"
        width="0"
        x="40"
        y="-16"
      />
    </g>
  </g>
</svg>
`;
